spring-integration-xml/src/main/java/org/springframework/integration/xml/DefaultXmlPayloadConverter.java

```diff
@@ -57,40 +61,63 @@ public DefaultXmlPayloadConverter(DocumentBuilderFactory documentBuilderFactory)
 
 	@Override
 	public Document convertToDocument(Object object) {
-		if (object instanceof Document) {
-			return (Document) object;
-		}
-		if (object instanceof Node) {
-			return nodeToDocument((Node) object);
-		}
-		else if (object instanceof DOMSource) {
-			Node node = ((DOMSource) object).getNode();
-			if (node instanceof Document) {
-				return (Document) node;
+		try {
+			if (object instanceof Document) {
+				return (Document) object;
 			}
-			else {
-				return nodeToDocument(node);
+			else if (object instanceof Node) {
+				return nodeToDocument((Node) object);
 			}
-		}
-		if (object instanceof File) {
-			try {
-				return getDocumentBuilder().parse((File) object);
+			else if (object instanceof DOMSource) {
+				Node node = ((DOMSource) object).getNode();
+				if (node instanceof Document) {
+					return (Document) node;
+				}
+				else {
+					return nodeToDocument(node);
+				}
 			}
-			catch (Exception e) {
-				throw new MessagingException("failed to parse File payload '" + object + "'", e);
+			else if (object instanceof Source) {
+				InputSource inputSource = sourceToInputSource((Source) object);
+				return getDocumentBuilder().parse(inputSource);
 			}
-		}
-		if (object instanceof String) {
-			try {
+			else if (object instanceof File) {
+				return getDocumentBuilder().parse((File) object);
+			}
+			else if (object instanceof String) {
 				return getDocumentBuilder().parse(new InputSource(new StringReader((String) object)));
 			}
-			catch (Exception e) {
-				throw new MessagingException("failed to parse String payload '" + object + "'", e);
+			else if (object instanceof InputStream) {
+				return getDocumentBuilder().parse((InputStream) object);
 			}
+			else if (object instanceof byte[]) {
+				return getDocumentBuilder().parse(new ByteArrayInputStream((byte[]) object));
+			}
+		}
+		catch (Exception e) {
+			throw new MessagingException("failed to parse " + object.getClass() + " payload '" + object + "'", e);
 		}
+
 		throw new MessagingException("unsupported payload type [" + object.getClass().getName() + "]");
 	}
 
+	private static InputSource sourceToInputSource(Source source) {
+		if (source instanceof SAXSource) {
+			return ((SAXSource) source).getInputSource();
+		}
+		else if (source instanceof StreamSource) {
+			StreamSource ss = (StreamSource) source;
+			InputSource inputSource = new InputSource(ss.getSystemId());
+			inputSource.setByteStream(ss.getInputStream());
+			inputSource.setCharacterStream(ss.getReader());
+			inputSource.setPublicId(ss.getPublicId());
+			return inputSource;
+		}
+		else {
+			return new InputSource(source.getSystemId());
```

Comment:
The code above seems to be a copy of `SAXSource.sourceToInputSource()` - why not just call that and do this if null is returned?